molecule/provisioners/libvirtprovisioner.py

```diff
@@ -0,0 +1,923 @@
+    def name(self):
+        return 'libvirt'
+
+    @property
+    def instances(self):
+        return self.molecule.config.config['libvirt']['instances']
+
+    @property
+    def default_provider(self):
+        return 'libvirt'
+
+    @property
+    def default_platform(self):
+        return ''
+
+    @property
+    def provider(self):
+        return self._provider
+
+    @property
+    def platform(self):
+        return self._platform
+
+    @property
+    def host_template(self):
+        return '{} ansible_ssh_host={} ansible_ssh_private_key_file={} ansible_ssh_user={}\n'
+
+    @property
+    def valid_providers(self):
+        return [{'name': 'Libvirt'}]
+
+    @property
+    def valid_platforms(self):
+        return [{'name': 'Libvirt'}]
+
+    @property
+    def ssh_config_file(self):
+        return '.molecule/ssh_config'
+
+    @property
+    def testinfra_args(self):
+        kwargs = {
+            'ansible-inventory':
+            self.molecule.config.config['ansible']['inventory_file'],
+            'connection': 'ansible'
+        }
+
+        return kwargs
+
+    @property
+    def serverspec_args(self):
+        return dict()
+
+    @property
+    def ansible_connection_params(self):
+        params = {'connection': 'ssh'}
+
+        return params
+
+    def _define_pool(self, pool_name):
+        """
+        Define a libvirt storage pool and start it.
+
+        :param: the name of the pool
+        :param: the filesystem path for the pool
+        """
+        utilities.print_info("Creating libvirt storage pool for molecule ...")
+        pool = ET.Element('pool', type='dir')
+        ET.SubElement(pool, 'name').text = pool_name
+        target = ET.SubElement(pool, 'target')
+        ET.SubElement(target, 'path').text = self._pool_path
+        poolxml = ET.tostring(pool)
+        LOG.debug("\tXMLDesc: {}".format(poolxml))
+        newpool = self._libvirt.storagePoolDefineXML(poolxml)
+        newpool.create()
+        return newpool
+
+    def _up_pool(self, pool_name='molecule'):
+        """
+        Up or define/up a libvirt storage pool
+
+        :param: name for the pool
+        """
+        try:
+            pool_found = self._libvirt.storagePoolLookupByName(pool_name)
+            if pool_found and not pool_found.isActive():
+                pool_found.create()
+            return pool_found  # existing pool is now running
+        except libvirt.libvirtError:
+            return self._define_pool(pool_name)
+
+    def _define_network(self, network=None):
+        """
+        Define a libvirt network and start it.
+        TODO: Make this all configurable. 'network' will be a dict that describes a libvirt network, populated based on molecule.yml.
+        libvirt:
+            networks:
+                - name: molecule0
+                  forward: nat|none
+                  #bridge: virbr10
+                  cidr: 192.168.121.1/24
+        """
+        cidr_net = netaddr.IPNetwork(network['cidr'])
```

Comment:
This `cidr_net` is only used much later in the code.  I believe you should define it just before you need it.  This will make the flow of the function easier to understand.